diff --git a/tests/basic.c b/tests/basic.c
index a0ad33e..95068cf 100644
--- a/tests/basic.c
+++ b/tests/basic.c
@@ -314,44 +314,30 @@ clutter_vector_registers(void)
   prepare_vector_data(data);
 
 #if defined(CLUTTER_VECTOR_REGISTER_AMD64)
-  asm volatile("movdqu %[data0], %%xmm0\n"
-	       "movdqu %[data1], %%xmm1\n"
-	       "movdqu %[data2], %%xmm2\n"
-	       "movdqu %[data3], %%xmm3\n"
-	       "movdqu %[data4], %%xmm4\n"
-	       "movdqu %[data5], %%xmm5\n"
-	       "movdqu %[data6], %%xmm6\n"
-	       "movdqu %[data7], %%xmm7\n"
-	       "movdqu %[data8], %%xmm8\n"
-	       "movdqu %[data9], %%xmm9\n"
-	       "movdqu %[data10], %%xmm10\n"
-	       "movdqu %[data11], %%xmm11\n"
-	       "movdqu %[data12], %%xmm12\n"
-	       "movdqu %[data13], %%xmm13\n"
-	       "movdqu %[data14], %%xmm14\n"
-	       "movdqu %[data15], %%xmm15\n"
+  asm volatile("leaq %[data], %%rax\n"
+	       "movdqu  (16 * 0)(%%rax), %%xmm0\n"
+	       "movdqu  (16 * 1)(%%rax), %%xmm1\n"
+	       "movdqu  (16 * 2)(%%rax), %%xmm2\n"
+	       "movdqu  (16 * 3)(%%rax), %%xmm3\n"
+	       "movdqu  (16 * 4)(%%rax), %%xmm4\n"
+	       "movdqu  (16 * 5)(%%rax), %%xmm5\n"
+	       "movdqu  (16 * 6)(%%rax), %%xmm6\n"
+	       "movdqu  (16 * 7)(%%rax), %%xmm7\n"
+	       "movdqu  (16 * 8)(%%rax), %%xmm8\n"
+	       "movdqu  (16 * 9)(%%rax), %%xmm9\n"
+	       "movdqu  (16 * 10)(%%rax), %%xmm10\n"
+	       "movdqu  (16 * 11)(%%rax), %%xmm11\n"
+	       "movdqu  (16 * 12)(%%rax), %%xmm12\n"
+	       "movdqu  (16 * 13)(%%rax), %%xmm13\n"
+	       "movdqu  (16 * 14)(%%rax), %%xmm14\n"
+	       "movdqu  (16 * 15)(%%rax), %%xmm15\n"
 	      :
-	      : [data0] "m" (*data[0]),
-	        [data1] "m" (*data[1]),
-	        [data2] "m" (*data[2]),
-	        [data3] "m" (*data[3]),
-	        [data4] "m" (*data[4]),
-	        [data5] "m" (*data[5]),
-	        [data6] "m" (*data[6]),
-	        [data7] "m" (*data[7]),
-	        [data8] "m" (*data[8]),
-	        [data9] "m" (*data[9]),
-	        [data10] "m" (*data[10]),
-	        [data11] "m" (*data[11]),
-	        [data12] "m" (*data[12]),
-	        [data13] "m" (*data[13]),
-	        [data14] "m" (*data[14]),
-	        [data15] "m" (*data[15])
+	      : [data] "m" (data)
 	      : "memory"
 #ifdef __SSE2__
 	       ,"xmm0", "xmm1", "xmm2", "xmm3", "xmm4", "xmm5", "xmm6", "xmm7",
 	        "xmm8", "xmm9", "xmm10", "xmm11", "xmm12", "xmm13", "xmm14",
-	        "xmm15"
+	        "xmm15", "rax"
 #endif
 	      );
 #elif defined(CLUTTER_VECTOR_REGISTER_I386)
